src/query.rs

```diff
@@ -65,6 +65,26 @@ impl Query {
         Ok(format!("Query({:?})", self.get()))
     }
 
+    pub(crate) fn __and__(&self, other: Query) -> Query {
+        let inner = tv::query::BooleanQuery::from(vec![
+            (tv::query::Occur::Must, self.inner.box_clone()),
```

Comment:
Could we check if `self` or `other` already is a `BooleanQuery` of all-`Must` and apend ourselves to that instead of building trees of `BooleanQuery`?

Reply:
Is there any reason why would you want this feature? Complicated syntaxes like `(query1 | query2) & query3` can be very troublesome to implement. Its also common to nest queries in Lucene.

Reply:
I fear for the reduced efficiency compared to having a flat list of subqueries. AFAIU, Tantivy will not optimize that query tree and its execution will directly mirror the structure of the input query.

Reply:
For our current definition,
```rust
pub(crate) struct Query {
    pub(crate) inner: Box<dyn tv::query::Query>,
}
```
the Query is actually accepting a dynamic dispatch `tantivy::query::Query` object. If we need to distinguish the presence of a boolean request, I afraid we need to setup dedicated Query struct but it will lead to big change in the API. The repackaging algorithm on the other hand, is not hard to deal with.

Reply:
>  If we need to distinguish the presence of a boolean request, I afraid we need to setup dedicated Query struct but it will lead to big change in the API. 

I was rather thinking about downcasting here. I also think we can keep this internal as an optimization instead of changing the API, i.e. downcast to see if we already have a boolean query with matching modifiers and append a clause instead of nesting.

Reply:
While I think the idea is fine, the optimization cannot be too deep due to the `Occur` implementation.

Currently we have 3 variants of `Occur`: `Should`, `Must`, `MustNot`. The optimization can be done when we have a tuple of `(Occur, BooleanQuery)` where the boolean query might contain one or more ***groups*** of `Should`, `Must`, `MustNot`.

Their mathematics goes like the symmetric matrix (Take the header as the index 0 `Occur` of the `(Occur, BooleanQuery)` tuple, while the rows `Occur` as the ***groups*** inside the `BooleanQuery`):

||`Should`|`Must`|`MustNot`|
|---|---|---|---|
|`Should`|`Should`|`(Must, [Should, ...])`|`(MustNot, [Should, ...])`|
|`Must`|`(Should, [Must, ...])`|`Must`|`MustNot`|
|`MustNot`|`(Should, [MustNot, ...])`|`MustNot`|`Must`|

This means that we can only ***flatten 1 layer*** those 5 types of `(Occur, InnerOccurGroup)`, am I right?

Reply:
> I also think we can keep this internal as an optimization instead of changing the API.

So do you mean you expect whenever `__and__` and `__or__` are used, the above optimization evaluation rule applies while the original `Query.boolean_query` remains the status quo?

Reply:
> So do you mean you expect whenever __and__ and __or__ are used, the above optimization evaluation rule applies while the original Query.boolean_query remains the status quo?

No, I just want `a & b & c` and `boolean_query([(Must, a), (Must, b), (Must, c)])` to produce the same internal representation using a single boolean query instead of the first form using two boolean queries.

Reply:
I think that's what I was going to convey in another way round.